backend.go

```diff
@@ -111,41 +129,82 @@ func Backend() *kubeAuthBackend {
 		InitializeFunc: b.initialize,
 	}
 
-	// Set default HTTP client
-	b.httpClient = cleanhttp.DefaultPooledClient()
-
-	// Set the review factory to default to calling into the kubernetes API.
-	b.reviewFactory = tokenReviewAPIFactory
-
 	return b
 }
 
 // initialize is used to handle the state of config values just after the K8s plugin has been mounted
 func (b *kubeAuthBackend) initialize(ctx context.Context, req *logical.InitializationRequest) error {
-	// Try to load the config on initialization
-	config, err := b.loadConfig(ctx, req.Storage)
+	config, err := b.config(ctx, req.Storage)
 	if err != nil {
 		return err
 	}
-	if config == nil {
-		return nil
+
+	if config != nil {
+		if err := b.updateTLSConfig(config); err != nil {
+			return err
+		}
 	}
 
-	b.l.Lock()
-	defer b.l.Unlock()
-	// If we have a CA cert build the TLSConfig
-	if len(config.CACert) > 0 {
-		certPool := x509.NewCertPool()
-		certPool.AppendCertsFromPEM([]byte(config.CACert))
-
-		tlsConfig := &tls.Config{
-			MinVersion: tls.VersionTLS12,
-			RootCAs:    certPool,
+	return b.runTLSConfigUpdater(context.Background(), req.Storage)
+}
+
+// runTLSConfigUpdater sets up a routine that periodically calls b.updateTLSConfig(). This ensures that the
+// httpClient's TLS configuration is consistent with the backend's stored configuration.
+func (b *kubeAuthBackend) runTLSConfigUpdater(ctx context.Context, s logical.Storage) error {
+	updateTLSConfig := func(ctx context.Context, s logical.Storage, force bool) error {
+		config, err := b.config(ctx, s)
+		if err != nil {
+			return fmt.Errorf("failed config read, err=%w", err)
+		}
+
+		if config == nil {
+			b.Logger().Trace("Skipping TLSConfig update, no configuration set")
+			return nil
+		}
+
+		if err := b.updateTLSConfig(config); err != nil {
+			return err
 		}
 
-		b.httpClient.Transport.(*http.Transport).TLSClientConfig = tlsConfig
+		return nil
 	}
 
+	horizon := time.Second * 30
+	ticker := time.NewTicker(horizon)
+	wCtx, cancel := context.WithCancel(ctx)
+	go func(ctx context.Context, cancel context.CancelFunc, s logical.Storage) {
+		sigs := make(chan os.Signal, 1)
+		signal.Notify(sigs, syscall.SIGHUP, syscall.SIGTERM, syscall.SIGINT)
```

Comment:
Interesting, I did test the signal handling for the built-in case and noticed no issues with Vault, aside from the fact that signal was not propagated down to the plugin. We would probably have to register each plugin's signal handler with the parent. Might be something worth considering as a feature to Vault.

I did not consider using fsnotify, but did was wary about pulling that in for systems where it may not be properly supported or configured.

We can probably forgo the signal handler setup, or make it optional, gated by env var or something?

Reply:
I did some more testing in with k8s-auth as a built-in, and it seems that the signals are being handled properly by vault and the plugin code. My previous test was probably against the external plugin case.

Reply:
In my experience it is almost always best to set up the inotify watcher on a directory, then match on files/subdirs and stats of interest.

Reply:
Yes I agree. But even there one could set the watch on directory where the files are (required in case of Kubernetes) or the parent directory (required if one would do atomic swap by renaming the whole parent directory). Wether watch works or not, completely depends on what update strategy the user has chosen.

Reply:
It feels like the most portable and safest solution would be to have Vault server handle the signal, and propagate it to all plugins.  That may be a fair bit of work.

Do you need to use a signal as the communication mechanism?  Could you add a reload endpoint instead?

Reply:
> It feels like the most portable and safest solution would be to have Vault server handle the signal, and propagate it to all plugins. That may be a fair bit of work.

Yeah, for the external plugin use case.

> 
> Do you need to use a signal as the communication mechanism? Could you add a reload endpoint instead?

Oh, yeah I like that idea even better! In the future, it could be interesting to allows plugins to register their own signal handlers with the Vault parent process.

Reply:
> In the future, it could be interesting to allows plugins to register their own signal handlers with the Vault parent process.

Maybe?  I'd rather we find other ways of communicating with users, signals aren't a great option for most uses cases IMO.  Lossy, no way for us to provide a response to the signaller, somewhat OS-specific...

Reply:
Dropped the signal handling bits with c9cfbdc8eeaac1bd951cb6531f59ba34880df7db